### PR TITLE
Fix handling for SELF, NEXT in doc_grammar

### PR DESCRIPTION
**Pull request by jfehrle**
The camlp5 documentation for SELF is inconsistent, which contributed to errors in how doc_grammar translated the construct.  I've checked representative examples in Coq to verify that it's correct now.

See https://camlp5.github.io/doc/pdf/camlp5-7.00.pdf version 7.00, June 26, 2017 section 8.6.2:

This seems correct but wasn't fully followed in the code: "When a SELF or the current entry name is encountered in the middle of the rule (i.e. if it is not the last symbol), there is a call to the "start" function of the First level of the current entry."

But this is incorrect, contradicting the first statement and leading to some confusion: "The left "SELF"s of the two levels "minus" and "power" correspond to a call to the next level."

Examples:
once idtac + (once idtac).  ACCEPTED -> lhs is top level
idtac + once idtac.  PARSE ERROR -> rhs is not top level, should be current level

`g_ltac.mlg` includes the following productions.  

```
  tactic_expr:
    | "3" RIGHTA [
      | IDENT "once"; ta = tactic_expr -> { TacOnce ta }
    | "2" RIGHTA
      | ta0 = tactic_expr; "+"; ta1 = tactic_expr -> { TacOr (ta0,ta1) }
```

You can see the fix for the example in `fullGrammar` at "tactic_expr2:".

### CODE REVIEW

**Comment by Zimmi48** [comment]
Thanks Jim!

Can I get an ack from a member of @coq/parsing-maintainers?

File: doc/tools/docgram/doc_grammar.ml [hunk 1]
```diff
@@ -417,21 +417,46 @@ let add_symdef nt file symdef_map =
...
+             (i.e. if it is not the last symbol), there is a call to the "start" function
+             of the First level of the current entry."
+
+            but this is incorrect, contradicting the first statement:
+            "The left "SELF"s of the two levels "minus" and "power" correspond to a call to the next level."
+
+            Examples:
+            once idtac + (once idtac).  ACCEPTED -> lhs goes to top
+            idtac + once idtac.  PARSE ERROR -> rhs is not to top level, sb at current level
+          *)
```
Comment (Zimmi48) [inline doc/tools/docgram/doc_grammar.ml:443]: This is what makes the linter complain. Should be squashed to make it happy.

```suggestion
          *)
```

**Comment by herbelin** [comment]
Hi Jim, I don't think we want to rely on this behavior of camlp5 which is only a "tolerance" and breaks the reversibility of the printer. See #9008 for a discussion. See also #11346 for an example of issues caused by a non-stratification of precedences in rules open on the left and/or on the right.

**Comment by jfehrle** [comment]
@herbelin Huh?  This only affects the appearance of the grammar in the documentation.  It is not built into the Coq executables.  This change makes doc_grammar follow the spec accurately and conforms to the observed behavior.

File: doc/tools/docgram/doc_grammar.ml (hunk 1 above)
Comment (jfehrle) [reply doc/tools/docgram/doc_grammar.ml]: The hook for removing whitespace runs whenever I commit.  I would have expected it would have fixed this.

**Comment by herbelin** [comment]
>  This change makes doc_grammar follow the spec accurately and conforms to the observed behavior.

I don't think that this should be the spec. The spec should be that parsing and printing are reverse, and this tolerance of camlp5 breaks the spec. The observed behavior should be considered a bug (and, if documented, documented as such). If it is not fixed, this is in large part for compatibility reasons. (Conversely, I tried at some time to see if the printer could be hacked to take into account this "tolerance" of camlp5 but I did not find a simple way to do it.)

File: doc/tools/docgram/doc_grammar.ml (hunk 1 above)
Comment (Zimmi48) [reply doc/tools/docgram/doc_grammar.ml]: Yeah, me too.

**Comment by jfehrle** [comment]
> I don't think that this should be the spec.

It's the only relevant spec I've seen; it was essential to me trying to understand our parsing (with almost no help despite asking).  I'm not aware of any parser doc specific to Coq.  I'd be happy to remove the references in the comments to the camlp5 document if you'd prefer that.

> The observed behavior should be considered a bug (and, if documented, documented as such). If it is not fixed, this is in large part for compatibility reasons.

Changing the current behavior could cause significant compatibility issues.  Maybe changing it is the right thing to do.  If the team agrees we should make those changes, that's fine with me.  But I think the team might want to see working code and an analysis of the impact on users beforehand.

I would need a description of the correct parser behavior in order to consider documenting the current behavior as a bug.  I actually haven't touched any documentation related to printing; I'm only describing the grammar.  And what would the doc say?  "While the parser accepts 'once idtac + (once idtac).' as a tactic, this is a bug, the correct form is to add parentheses like this: '(once idtac) + (once idtac)."?  I'm not sure such statements make Coq look good.

Let me know your thoughts.

**Comment by herbelin** [comment]
> Changing the current behavior could cause significant compatibility issues. Maybe changing it is the right thing to do. If the team agrees we should make those changes, that's fine with me. But I think the team might want to see working code and an analysis of the impact on users beforehand.

See https://github.com/herbelin/github-coq/commit/9b152597ea00c59b15f1e55ec37e564ab2f150e4 which shows a first excerpt of what would have to be changed. Most of the changes look to me as clarifying the expression. Note also that the changes correspond to what the printer would write.

> I would need a description of the correct parser behavior in order to consider documenting the current behavior as a bug.

The correct behavior is that:
- the left `SELF` of a right associative rule and the right `SELF` of a left associative rule are at next level
- the left `SELF` of a left associative rule and the right `SELF` of a right associative rule are at same level

>  "While the parser accepts 'once idtac + (once idtac).' as a tactic, this is a bug, the correct form is to add parentheses like this: '(once idtac) + (once idtac)."? I'm not sure such statements make Coq look good.

By the way, I did not notice at first, but the `once` parsing is here ok: `once idtac + (once idtac)` is parsed as `once (idtac + (once idtac))` which correctly follows the precedences. The case where precedences are not respected are typically the one at https://github.com/herbelin/github-coq/commit/9b152597ea00c59b15f1e55ec37e564ab2f150e4.

**Comment by herbelin** [comment]
PS: what may be counterintuitive though is that `once`, `progress`, `try` etc. are at level 3, while level 1 would have been (maybe) more intuitive (so that `once tac + once tac` is `(once tac) + (once tac)`). I'm afraid it is however too late for such a change.

**Comment by jfehrle** [comment]
How about submitting https://github.com/herbelin/github-coq/commit/9b152597ea00c59b15f1e55ec37e564ab2f150e4 in a draft PR and see how many tests and/or jobs it breaks?

It would be very helpful if there was a way to count how many lines of code would be affected if you make this change.  I suppose the change would cause some number of parse errors.  In other cases, code may still parse but would give different semantics, causing proof failures.

**Rambling off topic**: IIUC `coqc` currently quits upon finding first error.  It would be helpful if there was a command line option that could continue processing after the first error to the extent possible.  For example, resyncing the parser at the end of a proof or at the end of a sentence AND automatically treating any failing proofs as if they are Admitted (suppressing secondary errors AND reporting the total number of errors per file.  Those error counts would be give us a better idea of the impact of the change.  (I floated the auto-Admit part of this idea a while ago but no one saw the value then.  I also have stats-gathering code that could be repurposed to gather overall counts from a modified `coqc`  See https://github.com/coq/coq/pull/10652.)  **End rambling.**

With some data on the impact in hand, I'd suggest bringing it up for discussion online or in one of the weekly phone calls.  Or just submit as a PR if you prefer.

**Comment by Zimmi48** [comment]
> How about submitting [herbelin@9b15259](https://github.com/herbelin/github-coq/commit/9b152597ea00c59b15f1e55ec37e564ab2f150e4) in a draft PR and see how many tests and/or jobs it breaks?

I am taking bets on it breaking literally everything.



> > 
> 
> See [herbelin@9b15259](https://github.com/herbelin/github-coq/commit/9b152597ea00c59b15f1e55ec37e564ab2f150e4) which shows a first excerpt of what would have to be changed. Most of the changes look to me as clarifying the expression. Note also that the changes correspond to what the printer would write.

I'm not frankly convinced. In particular the `A /\ forall _, B` pattern is going to be super frequent and it's not clear to me why people would want to add parentheses.

Finally, I don't understand what bothers you with the fact that the printer will add a few extra-parentheses. People can decide to put more or fewer parentheses and that will not change the printer output (sometimes I wish it did). So what's the problem of having the printer output something that doesn't correspond to the minimally parenthesized expression?

**Comment by herbelin** [comment]
> Finally, I don't understand what bothers you with the fact that the printer will add a few extra-parentheses. People can decide to put more or fewer parentheses and that will not change the printer output (sometimes I wish it did). So what's the problem of having the printer output something that doesn't correspond to the minimally parenthesized expression?

I was in the remembrance that there were also cases where it is about missing parentheses, but maybe I'm wrong. Maybe is your pragmatic approach acceptable. I'll try to study the question further.

**Comment by Zimmi48** [comment]
> where it is about missing parentheses

That would be much more worrying indeed. It would be good to determine if that's the case or not. In the meantime, if this PR documents correctly the reality of the parser, I proposed to merge it.

**Comment by herbelin** [comment]
> In the meantime, if this PR documents correctly the reality of the parser, I proposed to merge it.

The changes made in the PR turn the grammar in an ambiguous one: if you have e.g.
```
forall A, A >: Prop
```
the modified grammar does not tell if this is `(forall A, A) : Prop` or `forall A, (A : Prop)`.

Also, the `once` example of the introduction is correct with respect to the current rules. There is nothing to change as far as I understand.

If you really want to document something, tell that there is a tolerance, in a remark, explaining the tolerance and telling that this tolerance is liable to evolve. But afaicj explaining this tolerance is not what this PR does.

**Comment by Zimmi48** [comment]
OK thanks! Now, I understand better your concerns with this PR and I agree that this tolerance could be documented in a note instead.

**Comment by jfehrle** [comment]
> By the way, I did not notice at first, but the `once` parsing is here ok: `once idtac + (once idtac)` is parsed as `once (idtac + (once idtac))` which correctly follows the precedences.

Yeah, I misinterpreted my example, which was the premise of the PR.  So I guess I'll drop the PR.  If you see any SELF/NEXT related problem with the old version of fullGrammar, please let me know.

**Pull request closed without merging**